### DISCUSSION TITLE
Compile error:   without an explicit x:DataType

### DISCUSSION THREAD

**Discussion by zy1075984** [AvaloniaUI/Avalonia#11952] [Q&A]
A compilation error occurs, but in another project, the same code has no error. Why?
11.0.0-rc1.1
axaml code:
                    <ContentControl Width="{Binding $parent.Bounds.Width}"
                                    MinWidth="400"
                                    MaxWidth="1200"
                                    HorizontalAlignment="Left"
                                    Content="{Binding ActiveElement}">

Compile error:  
       Cannot parse a compiled binding without an explicit x:DataType directive to give a starting data type for bindings

**Comment by zy1075984** [comment]
`<ContentControl Width="{Binding $parent.Bounds.Width}"
                                    MinWidth="400"
                                    MaxWidth="1200"
                                    HorizontalAlignment="Left"
                                    Content="{Binding ActiveElement}">`

**Comment by timunie** [comment]
Please read this carefully: https://docs.avaloniaui.net/docs/next/basics/data/data-binding/compiled-bindings

**Comment by timothyparez** [comment]
The documentation incorrectly states:

> Compiled bindings are not enabled by default. To enable compiled bindings, you will need to define the DataType of the object you want to bind to first. ...

I just created a new project using `dotnet new avalonia.mvvm` and `Compiled bindings` are enabled by default.
I had to explicitly disable them by adding `x:CompileBindings="True`  to the `XAML`  
  
Coming from `WPF` this confused me at first.

**Reply by workgroupengineering** [reply]
Open your csproj and set `AvaloniaUseCompiledBindingsByDefault` to `false`

**Reply by stevemonaco** [reply]
> The documentation incorrectly states:
> 
> > Compiled bindings are not enabled by default. To enable compiled bindings, you will need to define the DataType of the object you want to bind to first. ...

It's still technically correct as you need to opt into compiled bindings. The v11 project templates globally enable compiled bindings by default in the csproj (there is a creation option in the template to override this). Probably should have some clarification in the [v11 docs](https://docs.avaloniaui.net/docs/next/basics/data/data-binding/compiled-bindings#enable-and-disable-per-usercontrol-or-window). It's the better default as you should be using compiled bindings anyways. Prefer `{ReflectionBinding SomeProp}` for the few cases that aren't possible instead of disabling compiled bindings globally or within a XAML scope.

**Reply by timunie** [reply]
@timothyparez better to understand how it works, as compiled bindings have many benefits like performance and trimming friendly. Also better binding debugging. You'll love them after you understood how they work. 

Read this guide catefully: https://docs.avaloniaui.net/docs/next/basics/data/data-binding/compiled-bindings

**Reply by timothyparez** [reply]
I understand exactly how it works and I do think it's a good idea.
My issue is that

The documentation states:

> Compiled bindings are not enabled by default. To enable compiled bindings, you will need to define the DataType of the object you want to bind to first. ...

Which until, I read all the comments here, understood to mean that by default you are not required to define the `DataType`
when using `dotnet new avalonia.mvvm` unless you specifically enable it in the project by setting `AvaloniaUseCompiledBindingsByDefault` to true.

So in reality, compiled bindings ARE enabled by default and you ARE required to specify `DataType` by default.  
And in order to remove this requirement (which I agree you should not in most cases) you can set `AvaloniaUseCompiledBindingsByDefault` to false.

It would be easier to understand if the documentation said:

> Compiled bindings are enabled by default so you will need to define the DataType of the object you want to bind to first.

**Reply by timunie** [reply]
I just wanted to unblock you or otheres reading this thread. I agree that the docs need to reflect this changed behavior. But I need to understand what the v11 behavior is. Seems like it makes a difference how you create the project, so need to clarify this before fixing the docs in a wrong way 😆

**Reply by timothyparez** [reply]
If there's anything you need from me, let me know.

**Reply by timunie** [reply]
will do.
